### DISCUSSION TITLE
Concurrency for many HTTP requests

### DISCUSSION THREAD

**Discussion by ab2022** [cesanta/mongoose#2832] [Ask questions, get answers] [answered]
Hi,

I'm interested to use Mongoose as a web server that needs to handle many PUTs of small files to local disk from one client, and GETs of those same files from another client. The GETs are coming quickly, about 200 or 300 ms after the PUTs.

How would mongoose http server do concurrency in this case? Fork a process or spawn a thread per HTTP request? or is there a thread pool, or some pre-forked processes already running that distribute processing of the incoming requests? or some other way?

My platform is ubuntu 24.04 on AMD Ryzen, plenty of mem and cpu (16 cores @ 3 Ghz). 

Any info appreciated. Thanks.

**Comment by scaprile** [comment, marked as answer]
Please see our [documentation](https://mongoose.ws/documentation/),  and follow the guidelines in our [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples).
From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above) and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do [multi-threading](https://mongoose.ws/tutorials/core/multi-threaded/).

**Reply by scaprile** [reply]
In the very same doc linked above, you'll see the simplest web server.
